draft-irtf-cfrg-vdaf.md

```diff
@@ -1550,44 +1515,36 @@ struct {
 ~~~
 
 These messages are used to transition between the states described in
-{{vdaf-prep-comm}}. They are encoded and decoded to or from byte buffers as
-described {{Section 3 of !RFC8446}}) using the following routines:
+{{vdaf-prep-comm}}. The Leader's initial transition is computed with the
+following method, implemented on `Vdaf`:
 
-* `encode_ping_pong_message(message: Message) -> bytes` encodes a `Message` into
-  an opaque byte buffer.
-
-* `decode_pong_pong_message(encoded: bytes) -> Message` decodes an opaque byte
-  buffer into a `Message`, raising an error if the bytes are not a valid
-  encoding.
-
-The Leader's initial transition is computed with the following procedure:
-
-~~~ pseudocode
+~~~ python
 def ping_pong_leader_init(
-            vdaf,
-            vdaf_verify_key: bytes,
-            agg_param: bytes,
-            nonce: bytes,
-            public_share: bytes,
-            input_share: bytes,
-        ) -> tuple[State, bytes]:
+        self,
+        vdaf_verify_key: bytes,
+        agg_param: bytes,
+        nonce: bytes,
+        public_share: bytes,
+        input_share: bytes) -> tuple[State, Optional[bytes]]:
+    """Called by the leader to initialize ping-ponging."""
     try:
-        (prep_state, prep_share) = vdaf.prep_init(
+        (prep_state, prep_share) = self.prep_init(
             vdaf_verify_key,
             0,
-            vdaf.decode_agg_param(agg_param),
+            self.decode_agg_param(agg_param),
             nonce,
-            vdaf.decode_public_share(public_share),
-            vdaf.decode_input_share(0, input_share),
-        )
-        outbound = Message.initialize(
-            vdaf.encode_prep_share(prep_share))
-        return (
-            Continued(prep_state),
-            encode_ping_pong_message(outbound),
+            self.decode_public_share(public_share),
+            self.decode_input_share(0, input_share),
         )
+        encoded = self.encode_prep_share(prep_share)
+        outbound = bytes()
```

Comment:
Hmm, I'm exploring two options here, neither of which I think is very good.

Option 1: define encoder/decoders for each message type:
* encode_init_msg(prep_share)
* encode_cont_msg(prep_msg, prep_share)
* encode_finish_msg(prep_msg)

Option 2: a single function for encoding each message type:
```python3
```

Reply:
What I like about what we have currently is that it's explicit, relatively concise, and don't have a bunch of `if` statements.

Reply:
I like Option 1, but I think an even simpler approach for encoding would be a function that takes the "type" byte (initialize/continue/finish) & a sequence of byte strings, and encodes the type byte followed by each of the byte strings with prefixed length. This would be a very simple function (one loop, I suppose).

Decoding would look similar: it could be a function that returns the type byte & a sequence of byte strings (decoded as you would expect). Then the program logic could handle parsing the byte strings, and the byte wrangling would be deduplicated & hid behind a function.

What I don't like about the existing approach is that it duplicates the encoding/decoding logic in each location, and the encoding/decoding logic is inline with the "processing" logic, which makes it harder to follow. If this argument isn't compelling to you, I won't hold the PR up over it. WDYT?

Reply:
You're right, that's much nicer. Done.